packages/vm/tests/runner.ts

```diff
@@ -0,0 +1,63 @@
+const level = require('level')
+import { Transaction } from '@ethereumjs/tx'
+import VM from '../lib'
+import { SecureTrie as Trie } from '@ethereumjs/trie'
+import { DefaultStateManager } from '../lib/state'
+import Common from '@ethereumjs/common'
+import { Block } from '@ethereumjs/block'
+
+const main = async () => {
+  const txData = {
+    nonce: '0x1a',
+    gasPrice: '0xba43b7400',
+    gasLimit: '0x493e0',
+    to: '0x9ca228250f9d8f86c23690074c2b96d5f5479f79',
+    value: '0x0',
+    data:
+      '0xa9059cbb000000000000000000000000bfe465e7eb5a2928b5bf22bef93ad06089dc6179000000000000000000000000000000000000000000000000006a94d74f430000',
+    v: '0x1c',
+    r: '0xc04f048766bea3b20dcea6d2fcaecc92f1a615d337606ab6eec2a1e8e1f27bfe',
+    s: '0x1c57e62653724560b06c95be750d958840db9fa44ffa777b6625b344bfadbbcf',
+  }
+  const common = new Common({ chain: 'mainnet', hardfork: 'chainstart' })
+  const tx = Transaction.fromTxData(txData, { common })
+  const trie = new Trie(level('/Users/jochem/Library/Ethereum/ethereumjs/mainnet/state/'))
+
+  // Ensure we run on the right root
+  trie._setRoot(
+    Buffer.from('9dceb42e227ee6a244449d1d92cc5e0c17c63b520d0ff050d943baa8055d0ca6', 'hex')
+  )
+
+  const stateManager = new DefaultStateManager({ trie, common })
+  const vm = new VM({ stateManager, common })
+
+  vm.on('step', (o: any) => {
+    if (o.opcode.name == 'SSTORE') {
+      console.log(
+        o.stack[o.stack.length - 1].toString('hex'),
+        o.stack[o.stack.length - 2].toString('hex')
+      )
+    }
+  })
+
+  const block = Block.fromBlockData(
+    {
+      header: {
+        // Ensures our fees get added to coinbase after we run the Tx (this is important for the state root in the receipt)
+        coinbase: Buffer.from('52bc44d5378309ee2abf1539bf71de1b7d7be3b5', 'hex'),
+      },
+    },
+    { common }
+  )
```

Comment:
Whew. Sophisticated.

We might want to leave some generalized version of this script (with the changing parts as CL arguments or constants at the beginning of the file or something) + some more usages instructions as comments inside either the VM or client test folders.

Reply:
As a general comment on the tooling for debugging: it should be built around blocks, not transactions. If we have a block with multiple transactions then if a transaction like the third one fails we should first execute the earlier blocks to get the state right, then execute failing tx, and then the remaining transactions on top. Then finally we check the state root and other things like receipt trie and gas used.

Reply:
This is not around something specific, but these will be different debugging modules over time like in the `devp2p` library (`devp2p:rlpx`, `devp2p:peer` and so on), and each module will output the module specific things, so e.g. `vm:tx` the specifics on the tx run. Of course once can also activate this debug logger when running locks. I will also add a dedicated one for `runBlocks` (`vm:block`), just have started with the tx one. 😄  And doing all at once is just too much, so I will (we can) add over time. Will also give this some documentation mentioning the loggers available.